### PR TITLE
ProQuestFSG: Handle search phrases and booleans

### PR DESCRIPTION
**Pull request by maccabeelevine**
The ProQuest FSG API doesn't treat quotations marks as wrapping phrases, the way most APIs do.  This converts the query to match the API's expected CQL syntax.  It then has to deal with boolean operators submitted in the query.

### CODE REVIEW

**Comment by maccabeelevine** [comment]
This is way more complicated that it should be.  Booleans are handled fine by default (as far as I can see), but phrases are not, and adding the code to deal with phrases broke the usual boolean handling and required some more hacks.

Digging more into the manual, I also see that there are additional semantics related to matching against variants (i.e. common suffixes) that I don't think are handled correctly.  But it really comes down to what sorts of syntax we expect in the search field in the first place -- are the other backends expected to handle booleans, wildcards, and whatever else in the same way as Solr.  Does EDS?  Should ProQuest?  Etc.  After I have a better handle on that I can take a second look and try to build support for what we actually should be handling, and then I can ask @sturkel89 if/when you have time to help with testing.

I think advanced search may also have some problems, but not sure if they are reasonably solvable in the same PR.

**Comment by maccabeelevine** [comment]
I also see in the manual that there is a second format I could have used to send in queries, "MQL" or Mnemonic Query Language, instead of the CQL queries sent as part of SRU.  MQL seems to handle phrases the way we would expect, and it's also the native language of the ProQuest web applications, which presumably would make things easier to test.  But swapping this out would be a pretty big rewrite of the PQ integration, so not trying to tackle until I at least see how bad the situation is with CQL.

**Comment by demiankatz** [comment]
@maccabeelevine, there are no hard and fast requirements for how a search backend should behave, since backend behavior is often entirely beyond the control of VuFind (some APIs just do what they do, and we can't change that -- better to have a behaviorally inconsistent integration than none at all). That being said, Solr does tend to be the baseline expectation, so the closer we can get to that behavior in other backends, the less cognitive dissonance our users have to contend with. So I think your goals here seem reasonable.

Is that any help?

File: module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php [hunk 1]
```diff
@@ -152,11 +152,41 @@ protected function queryToString(Query $query)
     {
         // Clean and validate input:
         $index = $query->getHandler();
+        $queryString = $query->getString();
         if (empty($index)) {
             // No handler?  Just accept query string as-is; no modifications needed.
-            return $query->getString();
+            return $queryString;
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php]: Is there a reason this edge case doesn't get processed? I can understand why you wouldn't need to prefix with an index name in this situation, but is it still possible other parts of the string might need manipulation?

**Comment by maccabeelevine** [comment]
> @maccabeelevine, there are no hard and fast requirements for how a search backend should behave, since backend behavior is often entirely beyond the control of VuFind (some APIs just do what they do, and we can't change that -- better to have a behaviorally inconsistent integration than none at all). That being said, Solr does tend to be the baseline expectation, so the closer we can get to that behavior in other backends, the less cognitive dissonance our users have to contend with. So I think your goals here seem reasonable.
> 
> Is that any help?

Yes definitely.  I'd say that there's probably a significant gap between what this integration does right now vs. Solr syntax handling, and that's ok, but moving in the Solr direction would be a good thing.  So then the question becomes whether this is an incremental improvement.  Obviously phrase handling is a plus, but given the boolean hacking I had to do, I'm not 100% sure it's a net win.  Would appreciate @sturkel89 's perspective.

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/ProQuestFSG/QueryBuilderTest.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            ['(cql.serverChoice = "ice hockey")', 'ice AND hockey'],
+            ['(cql.serverChoice = "ice hockey")', 'ice and hockey'],
+            ['(cql.serverChoice adj "ice AND hockey")', '"ice AND hockey"'],
+
+            // OR to convert to the 'any' operator
+            ['(cql.serverChoice any "ice hockey")', 'ice OR hockey'],
+
+            // Booleans between phrases
+            ['(cql.serverChoice adj "ice hockey") AND (cql.serverChoice = "skate")', '"ice hockey" and skate'],
+            ['(cql.serverChoice adj "ice hockey") OR (cql.serverChoice any "skate")', '"ice hockey" or skate'],
```
Comment (demiankatz) [inline module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/ProQuestFSG/QueryBuilderTest.php:71]: I don't know how far down this rabbit hole you want to go, but do you need to add test cases for some crazier edge cases? What about nested quotes? Mismatched quotes? Nested booleans with parentheses? For example:

```
"ice hockey
"ice "hockey""
(ice or hockey) AND ("ice hockey" or skate)
```

**Review by demiankatz — CHANGES_REQUESTED** [review]
@maccabeelevine, some fairly naive questions below -- just trying to probe the edges of this problem, not to drive you insane. ;-)

File: module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php (hunk 1 above)
Comment (maccabeelevine) [reply module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php]: Oh yes, it's a problem!  The phrase fix below won't run if there's no defined handler.  The only bright side is that I can't actually get this pathway to execute -- it always picks up the default somewhere before it gets to this point.  So I figured at least this PR doesn't make things any worse. :)

Comment (demiankatz) [reply module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php]: Maybe worth a comment to that effect so we don't get confused in future. :-)

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/ProQuestFSG/QueryBuilderTest.php (hunk 2 above)
Comment (maccabeelevine) [reply module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/ProQuestFSG/QueryBuilderTest.php:71]: Yup, this is what I'm looking for input on -- to itemize whatever we think are some of the more reasonable options, and see if I can get them to work without breaking anything else.  This is a good starting point.

File: module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php [hunk 3]
```diff
@@ -152,11 +152,41 @@ protected function queryToString(Query $query)
     {
         // Clean and validate input:
         $index = $query->getHandler();
+        $queryString = $query->getString();
         if (empty($index)) {
-            // No handler?  Just accept query string as-is; no modifications needed.
-            return $query->getString();
+            // No handler? Should never get here. Use the default?
+            $index = 'cql.serverChoice';
```
Comment (maccabeelevine) [inline module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php:158]: This will at least lead to correct behavior (vs returning the queryString as is).  I would rather reference Options->defaultHandler directly, but I don't see a good way to access that here.  And again I don't see a way to get to this code pathway.

Comment (demiankatz) [reply module/VuFindSearch/src/VuFindSearch/Backend/ProQuestFSG/QueryBuilder.php:158]: This is intended to be lower-level code than the Options class (VuFind\Search depends on VuFindSearch, not the other way around), so I think this is a reasonable approach.
